src/Microsoft.Data/ArrowStringColumn.cs

```diff
@@ -0,0 +1,348 @@
+            int curArrayIndex = 0;
+            int numBuffers = _offsetsBuffers.Count;
+            while (curArrayIndex < numBuffers && rowIndex > _offsetsBuffers[curArrayIndex].Length - 1)
+            {
+                rowIndex -= _offsetsBuffers[curArrayIndex].Length - 1;
+                curArrayIndex++;
+            }
+            return curArrayIndex;
+        }
+
+        private ReadOnlySpan<byte> GetBytes(long index)
+        {
+            int offsetsBufferIndex = GetBufferIndexContainingRowIndex(ref index);
+            int currentOffset = _offsetsBuffers[offsetsBufferIndex].ReadOnlySpan[(int)index];
+            int nextOffset = _offsetsBuffers[offsetsBufferIndex].ReadOnlySpan[(int)index + 1];
+            int numberOfBytes = nextOffset - currentOffset;
+            return _dataBuffers[offsetsBufferIndex].ReadOnlySpan.Slice(currentOffset, numberOfBytes);
+        }
+
+        protected override object GetValue(long rowIndex)
+        {
+            var bytes = GetBytes(rowIndex);
+            unsafe
+            {
+                fixed (byte* data = &MemoryMarshal.GetReference(bytes))
+                    return _encoding.GetString(data, bytes.Length);
+            }
+        }
+
+        protected override object GetValue(long startIndex, int length)
+        {
+            var ret = new List<string>();
+            while (ret.Count < length)
+            {
+                ret.Add((string)GetValue((int)startIndex));
+            }
+            return ret;
+        }
+
+        protected override void SetValue(long rowIndex, object value)
+        {
+            throw new NotSupportedException(Strings.ImmutableStringColumn);
+        }
+
+        public new string this[long rowIndex]
+        {
+            get => (string)GetValue(rowIndex);
+            set => throw new NotSupportedException(Strings.ImmutableStringColumn);
+        }
+
+        public new List<string> this[long startIndex, int length]
+        {
+            get
+            {
+                var ret = new List<string>();
+                while (ret.Count < length)
+                {
+                    ret.Add((string)GetValue(startIndex++));
+                }
+                return ret;
+            }
+        }
+
+        public override Field Field => new Field(Name, StringType.Default, NullCount != 0);
+
+        public override int MaxRecordBatchLength(long startIndex)
+        {
+            if (Length == 0)
+                return 0;
+            int offsetsBufferIndex = GetBufferIndexContainingRowIndex(ref startIndex);
+            Debug.Assert(startIndex <= Int32.MaxValue);
+            return _offsetsBuffers[offsetsBufferIndex].Length - (int)startIndex;
+        }
+
+        private int GetNullCount(long startIndex, int numberOfRows)
+        {
+            int nullCount = 0;
+            for (long i = startIndex; i < numberOfRows; i++)
+            {
+                if (!IsValid(i))
```

Comment:
(not necessarily for this PR) - I wonder if we could speed this up by using PopCount: [https://docs.microsoft.com/en-us/dotnet/api/system.numerics.bitoperations.popcount?view=netcore-3.0](https://docs.microsoft.com/en-us/dotnet/api/system.numerics.bitoperations.popcount?view=netcore-3.0)

Reply:
Oh yeah, good idea. I didn't even know this existed. Filed this:
https://github.com/dotnet/corefxlab/issues/2704